### DISCUSSION TITLE
Programmatically cancel/abort cursors/range session

### DISCUSSION THREAD

**Discussion by LucHermitte** [neoclide/coc.nvim#4986] [Q&A] [answered]
Hello,

Is there a way to programmatically end a range session from Vim script code? 

I've seen we can feed `q`, but I'm looking for an equivalent of `call CocAction('runCommand', 'editor.action.addRanges', ...)` that removes all active cursors and terminates the current multi-cursors session.

**Comment by fannheyward** [comment]
No such feature yet.

**Reply by LucHermitte** [reply]
I'm not sure what is your workflow. Do you want me to open an issue about it?

**Reply by fannheyward** [reply]
What's your workflow of using cursor feature?

**Reply by LucHermitte** [reply]
Context: I'm looking to replace an [optional dependency](https://github.com/tomtom/stakeholders_vim) in my snippet & template engines -- that belongs to the first generation of template expander plugins. In particular I was looking at how I could use CoC multiple cursors on the jump-to-next-placeholder from my [lh-brackets plugin](https://github.com/LucHermitte/lh-brackets) (and the plugins that depend on it like mu-template)

The typical workflow is _hit-jump-next/prev-placeholder-key_, which selects the next/prev placeholder and enters SELECT-mode (by default). My idea was to search at that moment for all the other occurrences of the placeholder text, in order to register them on the fly as extra cursors. And to do so, I need to unregister currently active cursor ranges.

Given the already complex scripting of the jump-to-next-placeholder functions, my accumulated bad experience over things like `feedkey` and so on, I'm looking for a function to call (/command to execute), when CoC is detected, in order to deactivate current active ranges (if any) when jumping-to-the-next-placeholder.

Note: One of the particularity of lh-brackets/mu-template is that placeholders have extra-characters that appears in the buffer once the snippets/templates are expanded ([screencast](https://github.com/LucHermitte/lh-cpp/blob/master/doc/screencast-snippet-baseNV-class.gif)). Their ranges are not cached when they are expanded, they are searched at the last moment. Also, the snippet insertion and syntax are quite different for me to rely on CoC related features.

A `call CocAction('runCommand', 'editor.action.cancelRanges')` seemed to be the easier approach for everybody.

**Reply by fannheyward** [reply, marked as answer]
`editor.action.cancelRanges` is a good idea, will check it.

**Reply by LucHermitte** [reply]
Thanks!
